### ISSUE TITLE
opening_brace conflicts with implicit_return

### ISSUE THREAD

**Issue by ZevEisenberg** [realm/SwiftLint#3034]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

This seems related to https://github.com/realm/SwiftLint/issues/1728, but for implicit return in single-expression functions, which were added to Swift after that original bug was written.

### Describe the bug

The following code complains because of `implicit_return`, which I have enabled on purpose:

```swift
enum Foo {
    func foo() -> () -> Void {
        return {
            print("hello")
        }
    }
}
```

however, when I remove the `return`, then `opening_brace` complains on the opening brace of the closure:

```swift
enum Foo {
    func foo() -> () -> Void {
        {
            print("hello")
        }
    }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ echo "enum Foo {                                                       
    func foo() -> () -> Void {
        {
            print("hello")
        }
    }
}
" | Pods/SwiftLint/swiftlint lint --no-cache --use-stdin --enable-all-rules
Loading configuration from '.swiftlint.yml'
<nopath>:2:5: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
<nopath>:1:1: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
<nopath>:1:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly. (explicit_top_level_acl)
<nopath>:3:9: warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration. (opening_brace)
<nopath>:8:1: warning: Trailing Newline Violation: Files should have a single trailing newline. (trailing_newline)
Done linting! Found 5 violations, 0 serious in 1 file.
```

### Environment

* SwiftLint version: **0.38.2**
* Installation method used: **CocoaPods**
* Paste your configuration file:

```yml
disabled_rules:
- cyclomatic_complexity
- empty_parentheses_with_trailing_closure
- file_length
- function_body_length
- function_parameter_count
- large_tuple
- line_length
- nesting
- type_body_length
- identifier_name

opt_in_rules:
  - anyobject_protocol
  - array_init
  - closure_spacing
  - contains_over_filter_count
  - contains_over_filter_is_empty
  - contains_over_first_not_nil
  - contains_over_range_nil_comparison
  - convenience_type
  - empty_count
  - extension_access_modifier
  - fatal_error_message
  - first_where
  - identical_operands
  - implicit_return
  - last_where
  - legacy_random
  - modifier_order
  - multiline_arguments
  - operator_usage_whitespace
  - optional_enum_case_matching
  - prefer_self_type_over_type_of_self
  - redundant_objc_attribute
  - sorted_first_last
  - sorted_imports
  - static_operator
  - toggle_bool
  - unused_control_flow_label
  - unused_import
  - vertical_whitespace

type_name:
  min_length: 2

identifier_name:
    excluded:
        - URL

trailing_comma:
  mandatory_comma: true

statement_position:
  statement_mode: uncuddled_else

custom_rules:
  straight_quotes:
    included: "Localized.swift"
    name: "Straight Quotes"
    regex: "'|\""
    match_kinds:
      - docComment
    message: "Quotes and apostrophes in localized strings should be curly: “” ‘’"
    severity: warning

excluded:
  - Pods
  - GudTay/Resources/Generated
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **NO**
* Which Xcode version are you using (check `xcodebuild -version`)? **Xcode 11.3 (11C29) on macOS 10.15.2 (19C57)**
* Do you have a sample that shows the issue? **See above.**
